### PR TITLE
Minimize path conversions for chunk mmap and system index

### PR DESCRIPTION
**Pull request by JoeLoser**
###  :notebook_with_decorative_cover: Description

See commits one-by-one for the independent, isolated changes. This changes `chunk`'s `mmap` to work on `std::filesystem::path` instead of `vast::path`. When updating the callers in system index, that motivated the change of converting its internal uses of `vast::path` to be `std::filesystem::path`.

###  :memo: Checklist

- [ ] All user-facing changes have changelog entries.
- [ ] The changes are reflected on [docs.tenzir.com/vast](https://docs.tenzir.com/vast), if necessary.
- [ ] The PR description contains instructions for the reviewer, if necessary.

### :dart: Review Instructions

Commit-by-commit.

### CODE REVIEW

File: libvast/src/chunk.cpp
Comment (dominiklohmann) [inline libvast/src/chunk.cpp:65]:
```suggestion
      VAST_INFO("failed to get file size for filename {}: {}",
                            filename, err.message());
```

File: libvast/src/system/index.cpp
Comment (dominiklohmann) [inline libvast/src/system/index.cpp:180]: Still need to check `err` here.

File: libvast/src/system/index.cpp
Comment (dominiklohmann) [inline libvast/src/system/index.cpp:352]: Note: If you want to change these, you need to adapt the `active_partition_actor` type definition in `vast/system/actors.hpp`, and the corresponding lambdas in the `active_partition` function definition return value.

File: libvast/src/system/index.cpp
Comment (dominiklohmann) [inline libvast/src/system/index.cpp:587]: Same as above, except for the filesystem actor.

File: libvast/src/system/index.cpp
Comment (dominiklohmann) [inline libvast/src/system/index.cpp:867]: Error should also be checked here.

**Review by dominiklohmann — COMMENTED** [review]
Another good change. Just a few comments here.

File: libvast/src/system/index.cpp
Comment (JoeLoser) [reply libvast/src/system/index.cpp:180]: I'm not sure we do based on my reading at [CppRef](https://en.cppreference.com/w/cpp/filesystem/exists). I have checked `err` in previous uses of this function in this codebase, but after some more reading, I don't think it's required. My understanding is the function  will return `false` still but also populate `err` which could be used in log messages for example if desired.

File: libvast/src/system/index.cpp
Comment (JoeLoser) [reply libvast/src/system/index.cpp:867]: See https://github.com/tenzir/vast/pull/1504#discussion_r602732512

File: libvast/src/system/index.cpp
Comment (JoeLoser) [reply libvast/src/system/index.cpp:352]: Yeah, I went down that path for a bit this morning, but it seemed a bit more nuanced than that. I'd have to add `CAF_ADD_TYPE_ID` for `std::filesystem::path` and even when I did that and updated the callers appropriately (or so I think), there were still caf errors, which are pretty poor IMO. I'm hoping to change the actors last once all of the surrounding code doesn't use `vast::path` anymore. What do you think? It might be something good for us to pair on when we get to that point.

File: libvast/src/chunk.cpp
Comment (JoeLoser) [reply libvast/src/chunk.cpp]: Fixed, thanks.

File: libvast/src/system/index.cpp
Comment (dominiklohmann) [reply libvast/src/system/index.cpp:352]: The type ID part is the worst, and we struggle with its requirements all the time. We'll have to think about possibly forward-declaring `std::filesystem::path`, splitting our type ID blocks into multiple (again), or [adding a `caf::inspector_access<std::filesystem::path>` specialization](https://actor-framework.readthedocs.io/en/stable/TypeInspection.html#specializing-inspector-access).

After giving this some thought, I feel like we either pair as you suggest or we do it ourselves internally sometime over the coming weeks.

File: libvast/src/chunk.cpp
Comment (mavam) [inline libvast/src/chunk.cpp:64]:
```suggestion
      VAST_ERROR("failed to get file size for filename {}: {}", filename,
                 err.message());
```

File: libvast/src/system/index.cpp
Comment (dominiklohmann) [reply libvast/src/system/index.cpp:180]: I verified that you are indeed correct on this one.

File: libvast/src/system/index.cpp
Comment (dominiklohmann) [reply libvast/src/system/index.cpp:352]: I've implemented this in https://github.com/tenzir/vast/pull/1508

File: libvast/src/chunk.cpp
Comment (dominiklohmann) [reply libvast/src/chunk.cpp:64]: This is now the only blocker for this PR, I think.

I wonder whether it makes sense to change the return value of `chunk::mmap` to be `caf::expected<chunk_ptr>`, with the post-condition that if the return value is not an error, it is never nullptr. What do you think @mavam?

Comment (mavam) [reply libvast/src/chunk.cpp:64]: The general guideline should be: if a binary distinction between `nullptr` and correctly created object is not enough information to understand failed construction at the call site, then going through the `expected` monad is reasonable to enrich the error. In other words, if there a multiple causes failure and the call site wants to distinguish them, then `expected` makes sense.

Comment (dominiklohmann) [reply libvast/src/chunk.cpp:64]: I disagree. I personally prefer returning an error here because that can be forwarded to the client process so the user actually sees it.

Comment (mavam) [reply libvast/src/chunk.cpp:64]: My point is different: if you know exactly what failed, you can also generate the error at the call site. There is no value in `expected` in this case. If not, only then `expected` adds value.

Comment (dominiklohmann) [reply libvast/src/chunk.cpp:64]: Then printing here at all leads to duplicate error reporting.

File: libvast/src/system/index.cpp
Comment (JoeLoser) [reply libvast/src/system/index.cpp:352]: > I've implemented this in #1508

LGTM.  I'll try to incorporate it into the actors likely in a follow-up PR.

File: libvast/src/system/index.cpp
Comment (JoeLoser) [reply libvast/src/system/index.cpp:180]: > I verified that you are indeed correct on this one.

Thanks for double checking. I'll put up a PR that doesn't check the extra error in the one or two other call sites I've added it in form previous PRs.

File: libvast/src/chunk.cpp
Comment (mavam) [reply libvast/src/chunk.cpp]: Sorry, but now I disagree 🙂 . Let's apply the rationale by looking at the implementation:, which has 3 calls that can fail:

1. `file_size`
2. `open`
3. `mmap`

The call site doesn't know what failed when it receives `nullptr`. So we are in the second scenario that I described. Ergo: I advocate that we *do* need `expected<chunk_ptr>` here.

Comment (mavam) [reply libvast/src/chunk.cpp]: (As a result, there should not be logging in here because it will likely be done at the call site.)

Comment (JoeLoser) [reply libvast/src/chunk.cpp]: For now, I've changed it to `VAST_ERROR`. I think if we want to change the return value of this API, let's do that in a small, follow-up PR so it can be reviewed in isolation.

In general, should all callers of `std::filesystem::file_size` report an error if we can't get the file size?

FWIW, I don't think callers should be generating the error at the call site as that requires knowledge of the function implementation (e.g. did the file size check fail due to permissions, file not existing, etc...). If we want to go the `expected` route, it should carry the appropriate `err.message()` so the caller can consume it appropriately.

Comment (mavam) [reply libvast/src/chunk.cpp]: > For now, I've changed it to `VAST_ERROR`. I think if we want to change the return value of this API, let's do that in a small, follow-up PR so it can be reviewed in isolation.

Agreed.

> In general, should all callers of `std::filesystem::file_size` report an error if we can't get the file size?

Depends on the call site: some times it's a dealbreaker but sometimes it doesn't hurt. In practice we probably want to report this to `ERROR` or `WARNING` log levels.
 
> FWIW, I don't think callers should be generating the error at the call site as that requires knowledge of the function implementation (e.g. did the file size check fail due to permissions, file not existing, etc...). If we want to go the `expected` route, it should carry the appropriate `err.message()` so the caller can consume it appropriately.

I think we're saying the same thing here. The error handling at the call site is uniform, but you still want to know what *exactly* failed if there are multiple options. By consuming the error message, users get actionable insight in what went wrong.

Comment (JoeLoser) [reply libvast/src/chunk.cpp]: Yep, we're on the same page @mavam. I'll work on changing this API to return `expected` and introduce a follow-up PR on top of this soon.

Are we all good with this PR for now?

Comment (dominiklohmann) [reply libvast/src/chunk.cpp]: All good afaict. I'll let @mavam merge.

Comment (JoeLoser) [reply libvast/src/chunk.cpp]: See https://github.com/tenzir/vast/pull/1509 for the follow-up PR. I'll rebase it once this PR lands.

**Review by mavam — APPROVED** [review]
@dominiklohmann looked at the individual pieces already. I just looked at the `mmap` factory.

**Pull request merged**
